File: javascript/cluster_management/src/create_multi_region_clusters.js
import { DSQLClient, CreateClusterCommand, UpdateClusterCommand, waitUntilClusterActive } from "@aws-sdk/client-dsql";

export async function createMultiRegionCluster(region1, region2, witnessRegion) {

    const client1 = new DSQLClient({ region: region1 });
    const client2 = new DSQLClient({ region: region2 });

    const repo = process.env.GITHUB_REPOSITORY || "local";
    const runId = process.env.GITHUB_RUN_ID || "local";

    try {
        // We can only set the witness region for the first cluster
        console.log(`Creating cluster in ${region1}`);
        const createClusterCommand1 = new CreateClusterCommand({
            deletionProtectionEnabled: true,
            tags: {
                Name: "javascript multi region cluster 1",
                Repo: repo,
                Type: "cluster-management",
                RunId: runId
            },
            multiRegionProperties: {
                witnessRegion: witnessRegion
            }
        });

        const response1 = await client1.send(createClusterCommand1);
        console.log(`Created ${response1.arn}`);

        // For the second cluster we can set witness region and designate the first cluster as a peer
        console.log(`Creating cluster in ${region2}`);
        const createClusterCommand2 = new CreateClusterCommand({
            deletionProtectionEnabled: true,
            tags: {
                Name: "javascript multi region cluster 2",
                Repo: repo,
                Type: "cluster-management",
                RunId: runId
            },
            multiRegionProperties: {
                witnessRegion: witnessRegion,
                clusters: [response1.arn]
            }
        });

        const response2 = await client2.send(createClusterCommand2);
        console.log(`Created ${response2.arn}`);

        // Now that we know the second cluster arn we can set it as a peer of the first cluster
        const updateClusterCommand1 = new UpdateClusterCommand(
            {
                identifier: response1.identifier,
                multiRegionProperties: {
                    witnessRegion: witnessRegion,
                    clusters: [response2.arn]
                }
            }
        );

        await client1.send(updateClusterCommand1);
        console.log(`Added ${response2.arn} as a peer of ${response1.arn}`);

        // Now that multiRegionProperties is fully defined for both clusters
        // they'll begin the transition to ACTIVE
        console.log(`Waiting for cluster 1 ${response1.identifier} to become ACTIVE`);

        await waitUntilClusterActive(
            {
                client: client1,
                maxWaitTime: 300 // Wait for 5 minutes
            },
            {
                identifier: response1.identifier
            }
        );
        console.log(`Cluster 1 is now active`);

        console.log(`Waiting for cluster 2 ${response2.identifier} to become ACTIVE`);
        await waitUntilClusterActive(
            {
                client: client2,
                maxWaitTime: 300 // Wait for 5 minutes
            },
            {
                identifier: response2.identifier
            }
        );
        console.log(`Cluster 2 is now active`);
        console.log("The multi region clusters are now active");
        return { cluster1Id: response1.identifier, cluster2Id: response2.identifier };
    } catch (error) {
        console.error("Failed to create cluster: ", error.message);
        throw error;
    }
}

async function main() {
    const region1 = process.env.CLUSTER_1_REGION || "us-east-1";
    const region2 = process.env.CLUSTER_2_REGION || "us-east-2";
    const witnessRegion = process.env.WITNESS_REGION || "us-west-2";

    await createMultiRegionCluster(region1, region2, witnessRegion);
}

if (process.env.NODE_ENV !== 'test') {
    main();
}

File: java/cluster_management/src/main/java/org/example/CreateMultiRegionClusters.java
package org.example;

import java.time.Duration;
import java.util.List;
import java.util.Map;

import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.retries.api.BackoffStrategy;
import software.amazon.awssdk.services.dsql.DsqlClient;
import software.amazon.awssdk.services.dsql.DsqlClientBuilder;
import software.amazon.awssdk.services.dsql.model.CreateClusterRequest;
import software.amazon.awssdk.services.dsql.model.CreateClusterResponse;
import software.amazon.awssdk.services.dsql.model.GetClusterResponse;
import software.amazon.awssdk.services.dsql.model.UpdateClusterRequest;

public class CreateMultiRegionClusters {

    public static void main(String[] args) {
        Region region1 = Region.of(System.getenv().getOrDefault("CLUSTER_1_REGION", "us-east-1"));
        Region region2 = Region.of(System.getenv().getOrDefault("CLUSTER_2_REGION", "us-east-2"));
        Region witnessRegion = Region.of(System.getenv().getOrDefault("WITNESS_REGION", "us-west-2"));

        DsqlClientBuilder clientBuilder = DsqlClient.builder()
                .credentialsProvider(DefaultCredentialsProvider.builder().build());

        try (
            DsqlClient client1 = clientBuilder.region(region1).build();
            DsqlClient client2 = clientBuilder.region(region2).build()
        ) {
            List<GetClusterResponse> clusters = example(client1, client2, witnessRegion);
            System.out.println("Created clusters:");
            clusters.forEach(System.out::println);
        }
    }

    public static List<GetClusterResponse> example(DsqlClient client1, DsqlClient client2, Region witnessRegion) {
        String repo = System.getenv().getOrDefault("GITHUB_REPOSITORY", "local");
        String runId = System.getenv().getOrDefault("GITHUB_RUN_ID", "local");

        // We can only set the witness region on the first cluster.
        System.out.println("Creating cluster in " + client1.serviceClientConfiguration().region());
        CreateClusterRequest request1 = CreateClusterRequest.builder()
                .deletionProtectionEnabled(true)
                .multiRegionProperties(mrp -> mrp.witnessRegion(witnessRegion.toString()))
                .tags(Map.of(
                        "Name", "java multi region cluster",
                        "Repo", repo,
                        "Type", "cluster-management",
                        "RunId", runId
                ))
                .build();
        CreateClusterResponse cluster1 = client1.createCluster(request1);
        System.out.println("Created " + cluster1.arn());

        // For the second cluster we can set the witness region and designate
        // cluster1 as a peer.
        System.out.println("Creating cluster in " + client2.serviceClientConfiguration().region());
        CreateClusterRequest request2 = CreateClusterRequest.builder()
                .deletionProtectionEnabled(true)
                .multiRegionProperties(mrp ->
                        mrp.witnessRegion(witnessRegion.toString()).clusters(cluster1.arn())
                )
                .tags(Map.of(
                        "Name", "java multi region cluster",
                        "Repo", repo,
                        "Type", "cluster-management",
                        "RunId", runId
                ))
                .build();
        CreateClusterResponse cluster2 = client2.createCluster(request2);
        System.out.println("Created " + cluster2.arn());

        // Now that we know the cluster2 ARN we can set it as a peer of cluster1.
        UpdateClusterRequest updateReq = UpdateClusterRequest.builder()
                .identifier(cluster1.identifier())
                .multiRegionProperties(mrp ->
                        mrp.witnessRegion(witnessRegion.toString()).clusters(cluster2.arn())
                )
                .build();
        client1.updateCluster(updateReq);
        System.out.printf("Added %s as a peer of %s%n", cluster2.arn(), cluster1.arn());

        // MultiRegionProperties is fully defined so both clusters will begin the
        // transition to ACTIVE.
        System.out.printf("Waiting for cluster %s to become ACTIVE%n", cluster1.arn());
        GetClusterResponse activeCluster1 = client1.waiter().waitUntilClusterActive(
                getCluster -> getCluster.identifier(cluster1.identifier()),
                config -> config.backoffStrategyV2(
                        BackoffStrategy.fixedDelayWithoutJitter(Duration.ofSeconds(10))
                ).waitTimeout(Duration.ofMinutes(5))
        ).matched().response().orElseThrow();

        System.out.printf("Waiting for cluster %s to become ACTIVE%n", cluster2.arn());
        GetClusterResponse activeCluster2 = client2.waiter().waitUntilClusterActive(
                getCluster -> getCluster.identifier(cluster2.identifier()),
                config -> config.backoffStrategyV2(
                        BackoffStrategy.fixedDelayWithoutJitter(Duration.ofSeconds(10))
                ).waitTimeout(Duration.ofMinutes(5))
        ).matched().response().orElseThrow();

        return List.of(activeCluster1, activeCluster2);
    }
}
